transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,491 @@
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::crypto::{Crypto, CryptoConfig};
+use crate::muxer::QuicMuxer;
+use crate::{QuicConfig, QuicError};
+use futures::channel::{mpsc, oneshot};
+use futures::prelude::*;
+use quinn_proto::crypto::Session;
+use quinn_proto::generic::{ClientConfig, ServerConfig};
+use quinn_proto::{
+    ConnectionEvent, ConnectionHandle, DatagramEvent, EcnCodepoint, EndpointEvent, Transmit,
+};
+use std::collections::{HashMap, VecDeque};
+use std::io::IoSliceMut;
+use std::mem::MaybeUninit;
+use std::net::SocketAddr;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::Instant;
+use udp_socket::{RecvMeta, SocketType, UdpCapabilities, UdpSocket, BATCH_SIZE};
+
+/// Message sent to the endpoint background task.
+#[derive(Debug)]
+enum ToEndpoint<C: Crypto> {
+    /// Instructs the endpoint to start connecting to the given address.
+    Dial {
+        /// UDP address to connect to.
+        addr: SocketAddr,
+        /// The remotes public key.
+        public_key: C::PublicKey,
+        /// Channel to return the result of the dialing to.
+        tx: oneshot::Sender<Result<QuicMuxer<C>, QuicError>>,
+    },
+    /// Sent by a `quinn_proto` connection when the endpoint needs to process an event generated
+    /// by a connection. The event itself is opaque to us.
+    ConnectionEvent {
+        connection_id: ConnectionHandle,
+        event: EndpointEvent,
+    },
+    /// Instruct the endpoint to transmit a packet on its UDP socket.
+    Transmit(Transmit),
+}
+
+#[derive(Debug)]
+pub struct TransportChannel<C: Crypto> {
+    tx: mpsc::UnboundedSender<ToEndpoint<C>>,
+    rx: mpsc::Receiver<Result<QuicMuxer<C>, QuicError>>,
+    port: u16,
+    ty: SocketType,
+}
+
+impl<C: Crypto> TransportChannel<C> {
+    pub fn dial(
+        &mut self,
+        addr: SocketAddr,
+        public_key: C::PublicKey,
+    ) -> oneshot::Receiver<Result<QuicMuxer<C>, QuicError>> {
+        let (tx, rx) = oneshot::channel();
+        let msg = ToEndpoint::Dial {
+            addr,
+            public_key,
+            tx,
+        };
+        self.tx.unbounded_send(msg).expect("endpoint has crashed");
```

Comment:
https://github.com/mxinden/rust-libp2p/commit/1230178cd660783a57b895d377621abaa66346e7 showcases how to remove the bounded channels. Today it builds on top of https://github.com/libp2p/rust-libp2p/pull/2529, though it would as well work without.